### ISSUE TITLE
`AttributeError: 'Task' object has no attribute 'uncancel'. Did you mean: 'cancel'?` when the asyncio `loop.set_task_factory` factory returns a `asyncio.Future`-compatible object

### ISSUE THREAD

**Issue by graingert** [python/cpython#95097]
**Bug report**

https://docs.python.org/3.11/library/asyncio-eventloop.html#asyncio.loop.set_task_factory is documented as: "The callable must return a [asyncio.Future](https://docs.python.org/3.11/library/asyncio-future.html#asyncio.Future)-compatible object." however a number of apis now assume tasks have an `uncancel()` method:

```python
import asyncio
import signal
import os


class Task:
    def __init__(self, loop, coro, **kwargs):
        self._task = asyncio.Task(coro, loop=loop, **kwargs)

    def cancel(self, *args, **kwargs):
        return self._task.cancel(*args, **kwargs)

    def add_done_callback(self, *args, **kwargs):
        return self._task.add_done_callback(*args, **kwargs)

    def remove_done_callback(self, *args, **kwargs):
        return self._task.remove_done_callback(*args, **kwargs)

    @property
    def _asyncio_future_blocking(self):
        return self._task._asyncio_future_blocking

    def result(self, *args, **kwargs):
        return self._task.result(*args, **kwargs)

    def done(self, *args, **kwargs):
        return self._task.done(*args, **kwargs)

    def cancelled(self, *args, **kwargs):
        return self._task.cancelled(*args, **kwargs)

    def exception(self, *args, **kwargs):
        return self._task.exception(*args, **kwargs)

    def get_loop(self, *args, **kwargs):
        return self._task.get_loop(*args, **kwargs)


async def amain():
    os.kill(os.getpid(), signal.SIGINT)
    await asyncio.Event().wait()


class MyPolicy(asyncio.DefaultEventLoopPolicy):
    def new_event_loop(self, *args, **kwargs):
        loop = super().new_event_loop(*args, **kwargs)
        loop.set_task_factory(Task)
        return loop

asyncio.set_event_loop_policy(MyPolicy())
try:
    asyncio.run(amain())
except KeyboardInterrupt:
    print("ok!")
```

results in:
```
$ python3.10 demo.py
ok!
```

```pytb
$ python3.11 demo.py
Traceback (most recent call last):
  File "/usr/lib/python3.11/asyncio/runners.py", line 120, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/asyncio/base_events.py", line 650, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/home/graingert/demo.py", line 24, in result
    return self._task.result(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/graingert/demo.py", line 41, in amain
    await asyncio.Event().wait()
  File "/usr/lib/python3.11/asyncio/locks.py", line 213, in wait
    await fut
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/graingert/demo.py", line 52, in <module>
    asyncio.run(amain())
  File "/usr/lib/python3.11/asyncio/runners.py", line 187, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/asyncio/runners.py", line 122, in run
    if self._interrupt_count > 0 and task.uncancel() == 0:
                                     ^^^^^^^^^^^^^
AttributeError: 'Task' object has no attribute 'uncancel'. Did you mean: 'cancel'?
```

**Your environment**

- CPython versions tested on: 3.10, 3.11.0b4, a6daaf2a132efbb1965b4502ff8a8cf3b5afed0e
- Operating system and architecture:

**Comment by graingert** [comment]
@pablogsal @Fidget-Spinner tagging this is a release blocker for 3.11

**Comment by Fidget-Spinner** [comment]
A short summary of the issue and why it's a release blocker:

asyncio in 3.11 now calls `uncancel` on tasks. `uncancel` is a completely undocumented method which didn't exist in 3.10. This causes any custom task classes which don't define `uncancel` to be rejected by asyncio in 3.11.

**Comment by kumaraditya303** [comment]
> This is backwards incompatible behavior.

Strictly, it is not a backwards incompatible change, The docs mentions that "The callable must return a [asyncio.Future](https://docs.python.org/3.11/library/asyncio-future.html#asyncio.Future)-compatible object." so if the Task interface changes then you have to adapt your code to allow duck typing to work. See Yury's opinion on a similar issue https://github.com/python/cpython/issues/84979#issuecomment-1093873360. IMO a whatsnew entry would be sufficient here on porting to 3.11.

**Comment by pablogsal** [comment]
> so if the Task interface changes then you have to adapt your code to allow duck typing to work.

Unfortunately, that's not what backwards compatibility is defined. The interface itself changing is backwards incompatible if **existing** APIs now fail due to **extra** parameters added. This would be fine if the interface changes to add a new parameter that's defaulted to something and existing interfaces keep doing what they were doing but new ones don't.

**Comment by graingert** [comment]
Here's the original opinion that influenced set_name:

https://github.com/python/cpython/pull/8547#issuecomment-409245319

**Comment by pablogsal** [comment]
This should be fixed ASAP, because beta5 is tomorrow and we don't really want to delay this one.

**Comment by kumaraditya303** [comment]
> Unfortunately, that's not what backwards compatibility is defined. The interface itself changing is backwards incompatible if existing APIs now fail due to extra parameters added. This would be fine if the interface changes to add a new parameter that's defaulted to something and existing interfaces keep doing what they were doing but new ones don't.


Unfortunately, there are multiple new APIs like taskgroups and the timeout context manager which depend on the new `uncancel` method to work correctly and if the third party implementation does not provides it will not work.

**Comment by kumaraditya303** [comment]
The PR https://github.com/python/cpython/pull/95098 will break taskgroups and timeout context manager on third party task factories as it skips the calling the `uncancel` method if it does not exists.

**Comment by kumaraditya303** [comment]
I see two ways to fix this while preserving backwards compatibility with third party task factory implementations. 

- Monkey patch the task created by third party factory with a fallback `uncancel` implementation with a internal flag let's say `_cancel_requested` on the task.

- Add a fallback implementation to handle case when `uncancel` does not exists and still correctly work with third party implementations. (hard and may not work in all cases)

**Comment by pablogsal** [comment]
> Unfortunately, there are multiple new APIs like taskgroups and the timeout context manager which depend on the new `uncancel` method to work correctly and if the third party implementation does not provides it will not work.

New APIs that don't work without new stuff is ok. Old APIs that now don't work without the new stuff is not ok

**Comment by graingert** [comment]
> The PR #95098 will break taskgroups and timeout context manager on third party task factories as it skips the calling the `uncancel` method if it does not exists.

it doesn't break it - it just allows the CancelledError to propagate, which is the old api

**Comment by kumaraditya303** [comment]
I'll work on a PR to unblock 3.11 release with minimal changes as Pablo advised https://github.com/python/cpython/issues/95097#issuecomment-1193352673

**Comment by kumaraditya303** [comment]
See https://github.com/python/cpython/pull/95211

**Comment by ambv** [comment]
This is fixed in 3.11 and 3.12. Thanks, Thomas! ✨ 🍰 ✨

**Comment by gvanrossum** [comment]
Thanks, Kumar, too!
